platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/PullVirtualTreeRequest.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2021-2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.virtualmap.internal.reconnect;
+
+import com.swirlds.common.crypto.DigestType;
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.common.io.SelfSerializable;
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.virtualmap.internal.Path;
+import java.io.IOException;
+
+/**
+ * Used during the synchronization protocol to send data needed to reconstruct a single virtual node.
+ *
+ * <p>On the learner side, a request is created with a path and a hash in the old learner
+ * tree (if exists), then sent to the teacher. On the teacher side, requests are deserialized
+ * from the stream, and for every request a response is sent back to the learner.
+ */
+public class PullVirtualTreeRequest implements SelfSerializable {
+
+    private static final long CLASS_ID = 0xecfbef49a90334ffL;
+
+    private static class ClassVersion {
+        public static final int ORIGINAL = 1;
+    }
+
+    // Only used on the teacher side
+    private final VirtualTeacherTreeView teacherView;
+
+    // Only used on the learner side
+    private final VirtualLearnerTreeView learnerView;
```

Comment:
I agree, kind of. Here the existing pattern is replicated from "push" implementation for consistency. See `Lesson` and its related classes (e.g. `LeafDataLesson`), they all contain references to learner/teacher tree views.

Reply:
I see. But I don't find that pattern particularly great. Since this is a newly introduced code that might some day completely replace the old implementation, I suggest to introduce a new pattern that is easier to understand and maintain rather than replicate the existing ugly one.

Reply:
OK, let's deep dive into it.

`PullVirtualTreeResponse`. On the teacher side, when a response is written to the socket stream, it may contain leaf data (when the leaf is dirty). To get leaf data, a `RecordAccessor` object is needed. In the current version, this object is in the teacher tree view, this is why the response needs a reference to it. Instead, the `RecordAccessor` reference can be stored in `PullVirtualTreeResponse`, but it would be no better than a reference to `TeacherPullVirtualTreeView`. I tried to do all the work (e.g. load teacher hash) in the teacher task, create a response object and pass all necessary information to it, then send it to async output stream. The problem with this approach is for different nodes different information is written: reconnect state info, leaf data, etc. The response class delegates decision what to write to the teacher view.

On the learner side, I can probably avoid passing a reference to a learner view to the response. Instead of using `deserialize(inputStream, version)` method, some other method may be used, with a view passed in as an extra arg. I am not sure if it would be better than the current code. As with teacher side, it is the learner view that defines what to read from the stream.

`PullVirtualTreeRequest`. Requests appear to be simpler than responses. They are all the same (e.g. no extra info is sent for leaves vs internals), deserialization is also straightforward. The class is updated to not have references to teacher or learner views.